### PR TITLE
fix(cache): support multi-index and multiple queries per index

### PR DESCRIPTION
**Pull request by aymeric-giraudet**
**Summary**

#### [FX-2938](https://algolia.atlassian.net/browse/FX-2938)

Fixes part of #6258 and #6259 

**Result**

- A single index may cause up to 3 requests at once, if it has facets applied or hierarchical facets, so request params should be an array.
- Shape of a cached request is `{ method: 'search', args: [Array<QueryParams>] }`, not `{ method: 'search', args: Array<Array<QueryParams>> }` as it takes a single argument which is an array of params.

[FX-2938]: https://algolia.atlassian.net/browse/FX-2938?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Haroenv** [comment]
Which parts of https://github.com/algolia/instantsearch/issues/6258 and https://github.com/algolia/instantsearch/discussions/6259 aren't solved in this PR?

File: packages/instantsearch.js/src/types/results.ts [hunk 1]
```diff
@@ -105,7 +105,7 @@ type InitialResult = {
     params: RecommendParametersOptions['params'];
     results: RecommendResults['_rawResults'];
   };
-  requestParams?: SearchOptions;
+  requestParams?: SearchOptions[];
```
Comment (Haroenv) [inline packages/instantsearch.js/src/types/results.ts:108]: should this still be optional? maybe InitialResult as a whole optional but each key required?

**Comment by aymeric-giraudet** [comment]
@Haroenv this one : 

> Now click on Women, there is a data request(which handles SSR via the loader using getServerState) and there is also an unnecessary client side query.

We would need to react to `initialResults` changes and totally opt-out of client-side requesting, or just have Remix users not call the loader if possible as it may add unnecessary overhead

**Review by Haroenv — APPROVED** [review]

**Comment by Haroenv** [comment]
In that case I think the issue identified in https://github.com/algolia/instantsearch/issues/6258 can be marked as done, and the part where the remix integration docs is wrong can be a separate ticket :)

File: packages/instantsearch.js/src/types/results.ts (hunk 1 above)
Comment (aymeric-giraudet) [reply packages/instantsearch.js/src/types/results.ts:108]: Could be an empty array maybe, I still think it's undefined when there are no search widgets and only recommend ones though

**Review by dhayab — APPROVED** [review]
👏

**Pull request merged**